Semi lagrange fixes

lmoresi opened the pull request:
It took a while to find and neutralise a major advection scheme bug associated with particle ordering during parallel migration operations. This has been present a while and was masked by the projection operator I used to map semi-Lagrange variables to their underlying symbolic / meshVariable type.

I have fixed this, and am also removing the smoothing. The projection is probably not needed (for efficiency) but does mean that we can have our choice of arbitrary variables (continuous / discontinuous)  to track the advection so it can stay.

In reviewing, please check for left-over crap from the dismantling re-assembly.

This is the fix, but there is a lot of other code that I tried to clean up.

```python
with self._nswarm_psi.access():

                orig_index = self._nswarm_psi._nI0.data.copy().reshape(-1)

                with self.mesh.access(self._workVar):
                    self._workVar.data[
                        orig_index, :
                    ] = self._nswarm_psi.swarmVariable.data[:, :]
```

> Major bug (totally my fault) in which I failed to account for particles being reordered after sending and receiving. We now store the node and process for each particle which means we can take small substeps to the sample point and then snap back directly to the launch point.
> 
> This includes JC's swarm wormhole teleportation code which is required for long trajectories.
> 
> I'm turning off all smoothing for projection operators which was introduced to counter numerical problems which turned out to be related to this bug.
> 
> Also, added the possibility to use a sample swarm based on a discontinuous underlying variable. This gives a good spread of points and means that corners / facets where particles may be lost are avoided by particle paths.
>

lmoresi commented:
I am running the advection diffusion examples and they do seem to be doing the right thing. Tests are failing for several cases that check actual values of advection-diffusion solver. Needs some extra checks and some thought about the best sorts of tests for this.

julesghub commented on src/underworld3/meshing.py:
```diff
@@ -208,6 +208,8 @@ class boundary_normals_3D(Enum):
 
     def box_return_coords_to_bounds(coords):
 
+        epsilon = 1.0e-3
```
Why the need for epsilon here?

lmoresi replied:
We usually try to move points slightly inside boundaries because of the interpolation routines being a bit fussy. It should not matter with these straight boundaries and I am happy to eliminate this one.

lmoresi replied:
Removed

julesghub commented on src/underworld3/swarm.py:
```diff
@@ -1,3 +1,4 @@
+from posixpath import pardir
```
This doesn't appear to be used, right?

lmoresi replied:
Not sure how this made it into the swarm imports !

lmoresi replied:
Removed

lmoresi commented:
OK - so I am convinced that we have fixed the parallel transportation bug and I am now seeing good results with the Rayleigh-Bénard convection problem (constant viscosity). I have addressed @julesghub 's comments and I have reverted all unnecessary changes. I plan to merge with development so that these fixes can be evaluated in the complex benchmarks.
